model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/organization/InfinispanOrganizationProvider.java

```diff
@@ -151,24 +149,76 @@ public Stream<UserModel> getMembersStream(OrganizationModel organization, String
 
     @Override
     public UserModel getMemberById(OrganizationModel organization, String id) {
-        return orgDelegate.getMemberById(organization, id);
+        RealmModel realm = getRealm();
+        UserModel user = session.users().getUserById(realm, id);
+
+        if (user == null) {
+            return null;
+        }
+
+        String cacheKey = cacheKeyMembership(realm, organization, user);
+
+        if (isInvalid(cacheKey)) {
+            return orgDelegate.getMemberById(organization, user.getId());
+        }
+
+        CachedMembership cached = realmCache.getCache().get(cacheKey, CachedMembership.class);
+
+        if (cached == null) {
+            boolean isManaged = orgDelegate.isManagedMember(organization, user);
+            Long loaded = realmCache.getCache().getCurrentRevision(cacheKey);
+            UserModel member = orgDelegate.getMemberById(organization, user.getId());
+            cached = new CachedMembership(loaded, cacheKeyMembership(realm, organization, user), realm, isManaged, member != null);
+            realmCache.getCache().addRevisioned(cached, realmCache.getStartupRevision());
+        }
+
+        return cached.isMember() ? user : null;
     }
 
     @Override
-    public OrganizationModel getByMember(UserModel member) {
-        return orgDelegate.getByMember(member);
+    public Stream<OrganizationModel> getByMember(UserModel member) {
+        String cacheKey = cacheKeyByMember(member);
+
+        if (isInvalid(cacheKey)) {
+            return orgDelegate.getByMember(member);
+        }
+
+        CachedOrganizationIds cached = realmCache.getCache().get(cacheKey, CachedOrganizationIds.class);
+
+        if (cached == null) {
+            Long loaded = realmCache.getCache().getCurrentRevision(cacheKey);
+            Stream<OrganizationModel> model = orgDelegate.getByMember(member);
+            cached = new CachedOrganizationIds(loaded, cacheKey, getRealm(), model);
+            realmCache.getCache().addRevisioned(cached, realmCache.getStartupRevision());
+        }
+
+        return cached.getOrgIds().stream().map(this::getById);
     }
 
     @Override
-    public boolean isManagedMember(OrganizationModel organization, UserModel member) {
-        return orgDelegate.isManagedMember(organization, member);
+    public boolean isManagedMember(OrganizationModel organization, UserModel user) {
+        UserModel member = getMemberById(organization, user.getId());
+
+        if (member == null) {
+            return false;
+        }
```

Comment:
Can we rely on `CachedMembership`  instead of getting the member by id? I am probably missing something ... if CachedMembership is not invalidated then we may be able to get the information whether the users is Managed or not from it, don't we?

Reply:
If the `getMemberById` returns null is because the user is not a member.

Reply:
I see, so it means that user might not a member, but could still be cached the information within `CachedMembership`, right?